lib/Sema/AvailableFactsAnalysis.cpp

```diff
@@ -0,0 +1,499 @@
+  CollectExpressions(I.first, Exprs);
+  CollectExpressions(I.second, Exprs);
+  for (auto InnerExpr : Exprs)
+    if (const DeclRefExpr *DRE = dyn_cast<DeclRefExpr>(InnerExpr))
+      if (const VarDecl *VD = dyn_cast<VarDecl>(DRE->getDecl()))
+        if (VD == V)
+          return true;
+  return false;
+}
+
+// This function returns true only if expression `E` is a pointer deref.
+// A pointer deref is defined as follows:
+// - `E` should be a cast of type `LValueToRValue, and
+// - If the child of `E` is `e`, then `f(e)` must be true. `f(e)` is defined in
+//   a separate function `IsPointerDerefLValue`.
+// Otherwise, the function returns false;
+bool AvailableFactsAnalysis::IsPointerDeref(const Expr *E) {
+  if (const CastExpr *CE = dyn_cast<CastExpr>(E)) {
+    if (CE->getCastKind() != CastKind::CK_LValueToRValue)
+      return false;
+    return IsPointerDerefLValue(CE->getSubExpr());
+  }
+  return false;
+}
+
+bool AvailableFactsAnalysis::IsPointerAssignment(const Expr *E) {
+  if (const BinaryOperator *BO = dyn_cast<BinaryOperator>(E)) {
+    if (BO->isAssignmentOp())
+      if (IsPointerDerefLValue(BO->getLHS()))
+        return true;
+  }
+  if (const UnaryOperator *UO = dyn_cast<UnaryOperator>(E)) {
+    if (UO->isIncrementDecrementOp())
+      if (IsPointerDerefLValue(UO->getSubExpr()))
+        return true;
+  }
+  if (const CallExpr *CE = dyn_cast<CallExpr>(E)) {
+    for (int ArgIndex = 0; ArgIndex < CE->getNumArgs(); ArgIndex++) {
+      if (CE->getArg(ArgIndex)->getType()->isAnyPointerType())
+        return true;
+    }
+  }
+  return false;
+}
+
+// Given expression `E`, this function is defined as below:
+// 1. If `E` is one of the expressions (*, ->, or []) return true.
+// 2. If `E` is a parenthesized expression `(F)`, then return the result
+//    of this function on `F`.
+// 3. If `E` is a NoOp or LValueBitCast of `F`, then return the result
+//    of this function on `F`.
+// 4. Otherwise, return false.
+bool AvailableFactsAnalysis::IsPointerDerefLValue(const Expr *E) {
+  if (const UnaryOperator *UO = dyn_cast<UnaryOperator>(E))
+    if (UO->getOpcode() == UO_Deref)
+      return true;
+  if (const MemberExpr *ME = dyn_cast<MemberExpr>(E)) {
+    if (ME->isArrow())
+      return true;
+    else
+      return IsPointerDerefLValue(ME->getBase());
+  }
+  if (const ArraySubscriptExpr *AE = dyn_cast<ArraySubscriptExpr>(E))
+    return true;
+  if (const ParenExpr *PE = dyn_cast<ParenExpr>(E))
+    return IsPointerDerefLValue(PE->getSubExpr());
+  if (const CastExpr *CE = dyn_cast<CastExpr>(E))
+    if (CE->getCastKind() == CastKind::CK_LValueBitCast ||
+        CE->getCastKind() == CastKind::CK_NoOp)
+      return IsPointerDerefLValue(CE->getSubExpr());
+  return false;
+}
+
+// This function return true if an expression is volatile, and false otherwise.
+// An expression is volatile if there is at least one volatile variable in it.
+bool AvailableFactsAnalysis::IsVolatile(const Expr *E) {
```

Comment:
This is overly conservative: it would flag an expression that used &v, where v is a volatile variable, as being volatile.   I suggest that you use the existing function CheckIsNonModifying in the functions that are calling this instead.